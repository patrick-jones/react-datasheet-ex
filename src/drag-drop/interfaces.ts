import RDS from 'react-datasheet';
import {ConnectDragPreview, ConnectDragSource, ConnectDropTarget} from 'react-dnd';
import {ColumnHeader, HeaderRendererProps, RowRendererProps} from '../interfaces';

/**
 * Properties that must be provided to a drag and drop enabled row
 */
export interface DragDropRowProps<T extends RDS.Cell<T, V>, V = string> {
  /** The current row index */
  row: number;
  /** Callback for when a row is dropped on another */
  readonly onRowDropped: (source: number, target: number) => void;
}

/**
 * Properties that must be provided to a drag and drop enabled column header
 */
export interface DragDropHeaderProps {
  /** The current column index */
  col: number;
  /** Callback for one header is dropped on another */
  readonly onHeaderDropped: (source: number, target: number) => void;
}

/**
 * Drag source properties that are injected into a drag and drop provider
 */
export interface InjectedDragSourceProps {
  readonly connectDragSource: ConnectDragSource;
  readonly connectDragPreview: ConnectDragPreview;
}

/**
 * Drop target properties that are injected into a drag and drop provider
 */
export interface InjectedDropTargetProps {
  readonly connectDropTarget: ConnectDropTarget;
  readonly isDropOver: boolean;
}

/**
 * Combined drag source and drop target properties that are injected
 * into a drag and drop provider
 */
export interface InjectedDragDropProps extends InjectedDragSourceProps, InjectedDropTargetProps {}

/**
 * Extends `ReactDatasheet.RowRendererProps` to include
 * drag and drop properties
 */
export interface DragDropRowRendererProps<CellType, V = string>
  extends RowRendererProps<CellType, V>,
    DragDropRowProps<CellType, V> {}

/**
 * Extends `HeaderRendererProps` to include drag and drop properties.
 */
export interface DragDropHeaderRendererProps<T extends ColumnHeader = ColumnHeader>
  extends HeaderRendererProps<T>,
    DragDropHeaderProps {}
