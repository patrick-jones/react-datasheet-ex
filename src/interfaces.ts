import React, {CSSProperties} from 'react';
import RDS from 'react-datasheet';

export interface RowLocation {
  row: number;
}

export interface ColumnLocation {
  col: number;
}

export interface ColumnHeader {
  id: string;
  title: string;
  width?: number;
}

export interface HeaderRendererProps<T extends ColumnHeader = ColumnHeader> extends ColumnLocation {
  header: T;
}

export type HeaderRenderer<T extends ColumnHeader = ColumnHeader,
  R extends HeaderRendererProps<T> = HeaderRendererProps<T>> = React.ComponentType<R>;

export type HeaderContentRenderer<T extends ColumnHeader = ColumnHeader>
  = React.ComponentType<HeaderRendererProps<T>>;

export interface SheetRendererProps<
    T extends RDS.Cell<T, V>, V = string, H extends ColumnHeader = ColumnHeader
  >
  extends RDS.SheetRendererProps<T, V> {
  /** React style object that should be applied to the enclosing table */
  style?: CSSProperties | undefined;
  /**
   * An array of `ColumnHeader`, which are rendered into `th` elements
   * @default []
   */
  headers: H[];
  /**
   * Optional component or function to render the header cell contents.
   * The contents are rendered within a `th` element.
   * @default DefaultHeaderRenderer
   */
  headerRenderer: HeaderRenderer;
  /**
   * Optional component or function to render the header cell contents.
   * The contents are rendered within a `th` element.
   * @default DefaultHeaderContentRenderer
   */
  headerContentRenderer: HeaderContentRenderer;
}


export type ActionContentRenderer<T extends RowLocation = RowLocation>
  = React.ComponentType<T>;

export interface RowRendererProps<
    T extends RDS.Cell<T, V>, V = string, AP extends RowLocation = RowLocation
  > extends RDS.RowRendererProps<T, V> {

  actionContentRenderer?: ActionContentRenderer<AP>;
}
