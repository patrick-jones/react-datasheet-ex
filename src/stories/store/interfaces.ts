import RDS from 'react-datasheet';
import {ColumnHeader, Overflow} from '../../interfaces';

export type ExampleCellValue = string | number | null;
export interface ExampleCellType extends RDS.Cell<ExampleCellType> {
  value: ExampleCellValue;
}

export interface ExampleModel {
  id: string;
  headers: ExampleHeader[];
  rows: ExampleRow[];
  selected: RDS.Selection | null;
  overflow: Overflow;
}

export interface ExampleHeader extends ColumnHeader {
  hidden?: boolean;
}

export interface ExampleRow {
  id: string;
  selected?: boolean;
  data: ExampleCellType[];
}

