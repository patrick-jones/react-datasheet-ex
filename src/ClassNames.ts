export const READ_ONLY_CELL = 'cell read-only';
export const ACTION_CELL = 'rdx-action-cell';

export const HEADER_CELL = 'rdx-header-cell';
export const HEADER_CELL_LEFT = 'rdx-header-cell--left';
export const HEADER_CELL_CENTER = 'rdx-header-cell--center';
export const HEADER_CELL_RIGHT = 'rdx-header-cell--right';

export const HEADER_CELL_CONTAINER = 'rdx-header-cell-container';
export const HEADER_CELL_TITLE = 'rdx-header-cell-title';

export const HEADER_DRAG_SOURCE = 'rdx-header-drag-source';
export const HEADER_DROP_TARGET = 'rdx-header-drop-target';

export const ROW = 'rdx-row';
export const ROW_DROP_TARGET = 'rdx-row-drop-target';
export const ROW_DRAG_SOURCE = 'rdx-row-drag-source';

export const RESIZING_WIDTH = 'rdx-ex-resizing-width';
export const DATA_GRID_RESIZABLE = 'rdx-data-grid-resizable';
export const WIDTH_RESIZE_HANDLE = 'rdx-width-resize-handle';

export const SCROLL_PANEL = 'rds-ex-scroll-panel';
