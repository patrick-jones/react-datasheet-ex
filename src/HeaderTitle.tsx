import React, {ReactNode} from 'react';
import {ColumnHeader, HeaderRendererProps} from './interfaces';
import {HEADER_CELL_TITLE} from './ClassNames';
import headerDefaults from './headerDefaults';


interface HeaderTitleProps<T extends ColumnHeader = ColumnHeader> extends HeaderRendererProps<T> {
}


/**
 * A simple header content renderer that displays the header's `title` property.
 */
export default class HeaderTitle<T extends ColumnHeader = ColumnHeader>
  extends React.Component<HeaderTitleProps<T>> {

  static defaultProps = {
    className: '',
  };

  render(): ReactNode {
    const {overflow} = headerDefaults(this.props.header, {overflow: this.props.overflow});
    const {className, style} = this.props;

    const cn = [
      HEADER_CELL_TITLE,
      `${HEADER_CELL_TITLE}--${overflow}`,
      className,
    ].filter(c => !!c).join(' ');

    return (
      <span style={style} className={cn}>
        {this.props.children || this.props.header.title}
      </span>
    );
  }
}
