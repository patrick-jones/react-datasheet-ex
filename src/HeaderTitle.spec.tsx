import React from 'react';
import 'jest';
import {shallow} from 'enzyme';
import HeaderTitle from './HeaderTitle';
import {ColumnHeader} from './interfaces';
import {HEADER_CELL_TITLE} from './ClassNames';


describe('HeaderTitle', () => {
  const header: ColumnHeader = {
    id: 'abcd-puppy',
    title: 'AA',
  };

  it('Renders text as expected', () => {
    const wrapper = shallow(<HeaderTitle col={0} header={header} />);
    expect(wrapper.text()).toBe(header.title);
  });

  it('Renders custom text as expected', () => {
    const wrapper = shallow(<HeaderTitle col={0} header={header}>{header.id}</HeaderTitle>);
    expect(wrapper.text()).toBe(header.id);
  });

  it('renders overflow as expected', () => {
    const clip = shallow(<HeaderTitle col={0} header={header} />);
    expect(clip.find(`.${HEADER_CELL_TITLE}`).length).toBe(1);
    expect(clip.find(`.${HEADER_CELL_TITLE}--clip`).length).toBe(1);

    const wrap = shallow(<HeaderTitle col={0} header={header} overflow='wrap' />);
    expect(wrap.find(`.${HEADER_CELL_TITLE}`).length).toBe(1);
    expect(wrap.find(`.${HEADER_CELL_TITLE}--wrap`).length).toBe(1);
  });

});
