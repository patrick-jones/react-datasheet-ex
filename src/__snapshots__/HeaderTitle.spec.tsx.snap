// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HeaderTitle matches snapshot 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <HeaderTitle
    className=""
    col={0}
    header={
      Object {
        "id": "abcd-puppy",
        "title": "AA",
      }
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": "AA",
      "className": "rdx-header-cell-title rdx-header-cell-title--clip",
      "style": undefined,
    },
    "ref": null,
    "rendered": "AA",
    "type": "span",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": "AA",
        "className": "rdx-header-cell-title rdx-header-cell-title--clip",
        "style": undefined,
      },
      "ref": null,
      "rendered": "AA",
      "type": "span",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
